gitlab/client.py

```diff
@@ -1130,6 +1134,29 @@ def _query(
                 next_url = requests.utils.parse_header_links(result.headers["links"])[
                     0
                 ]["url"]
+            # if the next url is different with user provided server URL
+            # then give a warning it may because of misconfiguration
+            # but if the option to fix provided then just reconstruct it
+            if not next_url.startswith(self._gl.url):
+                search_api_url = re.search(r"(^.*?/api)", next_url)
+                if search_api_url:
+                    next_api_url = search_api_url.group(1)
+                    if self._gl.persist_base_url:
+                        next_url = next_url.replace(
+                            next_api_url, f"{self._gl._base_url}/api"
+                        )
+                    else:
+                        utils.warn(
+                            message=(
+                                f"The base url of the returned next page got "
+                                f"different with the user provided "
+                                f"{self._gl.url}/api* ~> {next_api_url}*, "
+                                f"since this may can lead to unexpected behaviour. "
+                                f"set argument persist_base_url to True for "
+                                f"resonctruct it with the origin one."
+                            ),
+                            category=UserWarning,
+                        )
```

Comment:
This gets us into quite a bit of nesting here and a big chunk inside the `try` block. We could probably avoid some with the use of `continue`, but if it's not feasible here just dismiss it and we can refactor later.

Reply:
this is the result of some warning from linter, it was shorter before. but yeah we can refactor it later.